cpp/src/join/mixed_join.cu

```diff
@@ -146,34 +148,33 @@ mixed_join(table_view const& left_equality,
 
   // If the join size data was not provided as an input, compute it here.
   std::size_t join_size;
-  std::unique_ptr<column> matches_per_row{};
-  column_view matches_view{};
+  // Using an optional because we only need to allocate a new vector if one was
+  // not passed as input, and rmm::device_uvector is not default constructible
+  std::optional<rmm::device_uvector<size_type>> matches_per_row{};
+  device_span<size_type> matches_per_row_span{};
 
   if (output_size_data.has_value()) {
-    join_size    = output_size_data->first;
-    matches_view = output_size_data->second;
+    join_size            = output_size_data->first;
+    matches_per_row_span = output_size_data->second;
   } else {
     // Allocate storage for the counter used to get the size of the join output
     rmm::device_scalar<std::size_t> size(0, stream, mr);
     CHECK_CUDA(stream.value());
 
-    matches_per_row = make_numeric_column(
-      data_type(type_to_id<size_type>()), outer_num_rows, mask_state::UNALLOCATED, stream, mr);
+    matches_per_row =
+      rmm::device_uvector<size_type>{static_cast<std::size_t>(outer_num_rows), stream, mr};
```

Comment:
Can't we call the `compute_mixed_join_output_size` function here instead of repeating the logic of calling the `compute_mixed_join_output_size` kernels? Then this logic could be simplified to an IILE like:
```
auto const& matches_per_row = output_size_data.has_value() ? output_size_data->second : compute_mixed_join_output_size(...);
```

Reply:
Ah, I guess that won't work because the types of the two results of the ternary aren't the same.

Reply:
Yeah you can't quite pull off a simple ternary like that. Also, the reason to use the kernels rather than the function is that the function does quite a bit of redundant setup that is already done inside the join function (generating table views, determining nullability, setting up all the variables, etc). This was briefly brought up in a review of the conditional join code back in https://github.com/rapidsai/cudf/pull/8928#discussion_r688245975. If we could find a way to reduce the redundancy between the size and the join kernels that would be awesome, but that's probably work for a future PR. Perhaps the solution is writing a detail API for the size computation that also handles all the common setup tasks for the higher-level size and join functions.